vltt/assignment3/solution.cpp

```diff
@@ -0,0 +1,197 @@
+        private:
+            std::map<char, Node*> go;
+        public:
+            bool terminal = false;
+            void add(const char &letter) {
+                go[letter] = new Node;
+            }
+            bool isHere(const char &letter) {
+                return go.find(letter) != go.end();
+            }
+            Node *next(const char &letter) {
+                if (go.find(letter) == go.end()) return nullptr;
+                return go[letter];
+            }
+        };
+        Node *start;
+    public:
+        Trie() {
+            start = new Node;
+        }
+        //add new word to the Trie
+        void add(const std::string &word) {
+            if (word.empty()) return;
+            Node *cur = start;
+            for (int i = 0; i < word.length(); i++) {
+                char letter = word[i];
+                if (cur->isHere(letter) == false) {
+                    cur->add(letter);
+                }
+                cur = cur->next(letter);
+                if (i == int(word.length()) - 1) {
+                    cur->terminal = true;
+                }
+            }
+        }
+        //determines whether there is such a prefix in the Trie
+        //return: -1 - prefix, 0 - nothing, 1 - word
+        int find_str(const std::string &str) {
+            if (str.empty()) return 0;
+            Node *cur = start;
+            for (int i = 0; i < str.length(); i++) {
+                if (cur == nullptr) break;
+                char letter = str[i];
+                cur = cur->next(letter);
+                if (i == int(str.length()) - 1) {
+                    if (cur == nullptr) return 0;
+                    if (cur->terminal == true) return 1;
+                    else return -1;
+                }
+            }
+            return 0;
+        }
+    };
+    Trie prefixes;
+public:
+    Dictionary() {}
+    Dictionary(std::vector<std::string> all_words) {
+        for (auto &word : all_words) {
+            add(word);
+        }
+    }
+    //add new word to the Dictionary
+    void add(const std::string &word) {
+        prefixes.add(word);
+    }
+    bool isWord(const std::string &cur) {
+        return prefixes.find_str(cur) == 1;
+    }
+    bool isPrefix(const std::string &cur) {
+        return prefixes.find_str(cur);
+    }
+};
+
+void passage_through_the_grid(const std::vector<std::vector<char> > &grid, Dictionary &dict, std::string &cur, int i_gr, int j_gr, std::set<std::string> &ans, std::set<std::pair<int, int> > &used) {
+    const std::vector<int> id_x = { 0, 1, -1 };
+    const std::vector<int> id_y = { 0, 1, -1 };
+    for (int i = 0; i < 3; i++) {
+        for (int j = 0; j < 3; j++) {
+            if (i == 0 && j == 0) continue;
+            int new_i = i_gr + id_x[i];
+            int new_j = j_gr + id_y[j];
+            std::pair<int, int> new_coordinates = std::make_pair(new_i, new_j);
+            if (new_i < 0 || new_i >= grid.size() || new_j < 0 || new_j >= grid[0].size()) continue;
```

Comment:
If you extract this into a helper function `in_bounds()` it would be easier to read. Alternatively you can add a comment to say that this if statement checks whether we can assess the node, so that whoever's reading this code doesn't have to read too much into it to figure out what it does.